mailpoet/assets/js/src/automation/integrations/mailpoet/components/form-token-field/index.tsx

```diff
@@ -0,0 +1,81 @@
+import { FormTokenField as WpFormTokenField } from '@wordpress/components';
+import { uniq } from 'lodash';
+
+export type FormTokenItem = {
+  id: number | string;
+  name: string;
+};
+
+export type FormTokenFieldProps = Omit<
+  WpFormTokenField.Props,
+  'value' | 'suggestions' | 'onChange'
+> & {
+  selected: FormTokenItem[];
+  suggestions: FormTokenItem[];
+  onChange: (values: FormTokenItem[]) => void;
+  label: string;
+  anyValue?: FormTokenItem;
+  anyValueIsDefault?: boolean;
+};
+
+export function FormTokenField({
+  label,
+  selected,
+  suggestions,
+  onChange,
+  anyValue,
+  anyValueIsDefault,
+  ...props
+}: FormTokenFieldProps): JSX.Element {
+  if (anyValue) {
+    suggestions.push(anyValue);
+  }
+  const uniqueSuggestions = uniq(suggestions);
+  return (
+    <WpFormTokenField
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      // The following error seems to be a mismatch. It claims the 'label' prop does not exist, but it does.
+      label={label}
+      value={selected.map((item) => item.name)}
+      suggestions={uniqueSuggestions.map((item) => item.name)}
+      onChange={(raw: string[]) => {
+        const allSelected: FormTokenItem[] = raw.map((item) => {
+          const match = uniqueSuggestions.find(
+            (suggestion) =>
+              suggestion.name.toLowerCase() === item.toLowerCase(),
+          );
+          return match;
+        });
+
+        // only the newly selected items
+        const newSelected = allSelected.filter((item) => {
+          for (let i = 0; i < selected.length; i += 1) {
+            if (selected[i].id === item.id) {
+              return false;
+            }
+          }
+          return true;
+        });
+        if (anyValue) {
+          const hasAny =
+            newSelected.filter((item) => item.id === anyValue.id).length > 0;
+          const hasNone = allSelected.length === 0 && anyValueIsDefault;
+          if (hasAny || hasNone) {
+            // when "any item" has been selected, all other items are removed
+            // or when no item has been selected but "any item" is default, we store "any item"
+            onChange([anyValue]);
+            return;
+          }
+        }
```

Comment:
Due to the [IF on line 30](https://github.com/mailpoet/mailpoet/pull/4371/commits/f4bde0a02b411648bace63f16310137e781d1863#diff-c993f976a5f0c987562cdc6ff34526cbe7d3d6fcc780207fbb204dd172528237R30), this whole block is unnecessary. It will work the same without it, I think.

And that's good — you could just ignore `anyItem` in the `onChange` callback, make it super-simple, and just watch the `selected` array with an effect such as:

```ts
useEffect(() => {
  if (selected.length === 0) {
    setSelected([...selected, anyValue]);
  }
  if (selected.contains(anyValue) && selected.length > 1) {
    setSelected([...selected.filter({ id }) !== anyValue.id]);
  }
}, [selected]);
```

(In this pseudocode I'm using `setSelected` but I mean the correct dispatch...)

Reply:
Problem outdated thanks to [710190b](https://github.com/mailpoet/mailpoet/pull/4371/commits/710190bd68357719381a9d577bc886a495f5a367)